From cc0c00591d64d964cc7f992f702329e53603f14d Mon Sep 17 00:00:00 2001
From: Aishwarya R <quic_aisr@quicinc.com>
Date: Mon, 11 Apr 2022 12:01:02 +0530
Subject: [PATCH] ath12k: discard BSS_CHANGED_TXPOWER when EXT_TPC_REG_SUPPORT
 for 6 GHZ

When STATION is connected to a 6 GHz AP, or when AP boots up, it has
2 way to configure the power limit to firmware. The first way is to
send 2 wmi command WMI_PDEV_PARAM_TXPOWER_LIMIT2G/
WMI_PDEV_PARAM_TXPOWER_LIMIT5G to firmware, the second way is to send
WMI_VDEV_SET_TPC_POWER_CMDID to firmware which include more parameters
for power control.

When firmware support SERVICE_EXT_TPC_REG, it means firmware support
the second way for WMI_VDEV_SET_TPC_POWER_CMDID, then ath12k discard
BSS_CHANGED_TXPOWER flag from mac80211 which is used to the first way
for 6 GHZ band in this patch and select the second way in the subsequent
patch.

Signed-off-by: Aishwarya R <quic_aisr@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 14 ++++++++++++--
 drivers/net/wireless/ath/ath12k/wmi.h |  1 +
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 9de6754..d61e5e2 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -3674,8 +3674,18 @@ static void ath12k_mac_op_bss_info_changed(struct ieee80211_hw *hw,
 		ath12k_dbg(ar->ab, ATH12K_DBG_MAC, "mac vdev_id %i txpower %d\n",
 			   arvif->vdev_id, info->txpower);
 
-		arvif->txpower = info->txpower;
-		ath12k_mac_txpower_recalc(ar);
+		if (ar->supports_6ghz && info->chandef.chan &&
+                    info->chandef.chan->band == NL80211_BAND_6GHZ &&
+                    (arvif->vdev_type == WMI_VDEV_TYPE_STA ||
+                     arvif->vdev_type == WMI_VDEV_TYPE_AP) &&
+                    test_bit(WMI_TLV_SERVICE_EXT_TPC_REG_SUPPORT,
+                             ar->ab->wmi_ab.svc_map)) {
+                        ath12k_dbg(ar->ab, ATH12K_DBG_MAC,
+                                   "discard tx power, change to set TPC power\n");
+                } else {
+                        arvif->txpower = info->txpower;
+                        ath12k_mac_txpower_recalc(ar);
+                }
 	}
 
 	if (changed & BSS_CHANGED_MCAST_RATE &&
diff --git a/drivers/net/wireless/ath/ath12k/wmi.h b/drivers/net/wireless/ath/ath12k/wmi.h
index af7c2f9..8113ec4 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -2108,6 +2108,7 @@ enum wmi_tlv_service {
 	WMI_TLV_SERVICE_MBSS_PARAM_IN_VDEV_START_SUPPORT = 253,
 	WMI_MAX_EXT_SERVICE = 256,
 
+	WMI_TLV_SERVICE_EXT_TPC_REG_SUPPORT = 280,
 	WMI_TLV_SERVICE_REG_CC_EXT_EVENT_SUPPORT = 281,
 
 	WMI_TLV_SERVICE_BE = 289,
-- 
2.17.1

